Add GPU kernel for SparseFillEmptyRows[Grad]

benbarsdell opened the pull request:
Follow-up to https://github.com/tensorflow/tensorflow/pull/44719

cc @nluehr

benbarsdell committed:
Add GPU kernel for SparseFillEmptyRows

benbarsdell committed:
Add GPU kernel for SparseFillEmptyRowsGrad

benbarsdell committed:
Change SparseFillEmptyRows GPU to an AsyncOpKernel

- Uses ThenExecute() to avoid stalling execution while waiting for data
  to be copied from device to host.

benbarsdell committed:
Add missing `return`s to SparseFillEmptyRows GPU

benbarsdell committed:
Fix RadixSort issue in SparseFillEmptyRows GPU

- gpuprim::DeviceRadixSort::SortPairs does not support in-place
  operation, so the input and output tensors must be different.

benbarsdell committed:
Move CHECK(done) up in SparseFillEmptyRows GPU

benbarsdell committed:
Enable sparse_fill_empty_rows tests on GPU

benbarsdell committed:
Fix two minor issues in SparseFillEmptyRows GPU

- Don't register the grad kernel for bool type.
- Include a header to ensure that GPU complex operators are available.

benbarsdell commented:
I fixed a few small issues I noticed since submitting this. I'm not sure why the Windows GPU build is complaining about std::complex on the device.

benbarsdell committed:
Share code b/w CPU and GPU for SparseFillEmptyRows

benbarsdell committed:
Fix some minor issues in SparseFillEmptyRows

- Fix copyright year.
- Remove use_gpu=True, which is already the default.
- Fix inline parameter name comment syntax.
- Allocate (some) outputs by name instead of index.

benbarsdell committed:
Report invalid indexes in SparseFillEmptyRows GPU

- Also adds a test to check that the error is raised.

benbarsdell commented on tensorflow/core/kernels/sparse_fill_empty_rows_op.cc:
```diff
@@ -236,6 +237,92 @@ TF_CALL_ALL_TYPES(REGISTER_CPU_KERNELS);
+        context, TensorShapeUtils::IsScalar(default_value_t.shape()),
+        errors::InvalidArgument("default_value must be a scalar, saw: ",
+                                default_value_t.shape().DebugString()),
+        done);
```
Done.

benbarsdell commented on tensorflow/python/kernel_tests/sparse_ops_test.py:
```diff
@@ -520,7 +520,7 @@ def testFillNumber(self):
   @test_util.run_deprecated_v1
   def testFillFloat(self):
-    with self.session(use_gpu=False):
+    with self.session(use_gpu=True):
```
Done.

benbarsdell commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,637 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
Done.

benbarsdell commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,637 @@
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+#if GOOGLE_CUDA
+#include "tensorflow/core/util/cuda_solvers.h"  // For ScratchSpace
```
Good idea, I'll do this after landing this PR.

benbarsdell commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,637 @@
+    // 5) Synchronize and allocate outputs (the sync is done implicitly by
+    //    enqueueing the remainder of the computation onto the stream as a host
+    //    callback).
+    // 6) If rows are not ordered:
```
The doc says `The output SparseTensor will be in row-major order` and I believe this behavior is relied-upon by some higher-level APIs.

sanjoy replied:
I believe "row major" in that context means that if the logical shape of the output is (say) `[A, B, C]` then the first element of an index corresponds to `A`, second one corresponds to `B` etc.

benbarsdell replied:
My understanding is that it is supposed to behave like tf.sparse.reorder, which is documented as "Reorders a SparseTensor into the canonical, row-major ordering."

There is a test for it: https://github.com/tensorflow/tensorflow/blob/4f80ad3/tensorflow/python/kernel_tests/sparse_ops_test.py#L588

The CPU implementation also explicitly checks if the rows are ordered, and only early-exits if they are: https://github.com/tensorflow/tensorflow/blob/69967c0/tensorflow/core/kernels/sparse_fill_empty_rows_op.cc#L129-L136

penpornk replied:
I agree with @benbarsdell. The behavior here is consistent with the CPU kernel:
* Can take 2D SparseTensors with unordered tuples.
* Returns a 2D SparseTensor with nonzeroes stored in row-major order.

sanjoy replied:
SG, I'll proceed with the rest of the review.

benbarsdell commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,637 @@
+      TensorShape output_indices_shape({N_full, rank});
+      OP_REQUIRES_OK_ASYNC(
+          context,
+          context->allocate_output(kOutputIndicesOutput, output_indices_shape,
```
Done for the indices and values outputs. The other two require using `output_required()` which only accepts indexes, so I've left them as-is.

benbarsdell commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,637 @@
+          /* d_in */ empty_row_indicator,
+          /* d_out */ num_empty_rows_through.data(),
+          /* num_items */ dense_rows,
+          /* stream */ cu_stream);
```
Done.

benbarsdell commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,637 @@
+    Tindex row = indices[i * rank];
+    if (row < 0 || row >= dense_rows) {
+      continue;  // Ignore indices that are out of range
+    }
```
After looking at this I realized that we can report invalid indexes like the CPU does without adding much overhead, so I've done that and also added a test that the error gets raised.

benbarsdell commented:
(Fixed Ubuntu CI issues).

benbarsdell committed:
Add ScopedActivateExecutorContext in async lambda

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op.cc:
```diff
@@ -185,41 +188,63 @@ struct SparseFillEmptyRows<CPUDevice, T, Tindex> {
+template <typename Device, typename T, typename Tindex>
+void SparseFillEmptyRowsOpImpl(
+    OpKernelContext* context, AsyncOpKernel::DoneCallback done = nullptr) {
+  if (!done) {
+    done = [] {};
+  }
```
This could be a default value for `done` right?

benbarsdell replied:
I actually did that originally but the CI failed with some bizarre compiler/linker errors. I couldn't reproduce it myself but guessed that this was the problem. I've added a comment to that effect.

sanjoy replied:
SG

benbarsdell committed:
Add comment re lambda in SparseFillEmptyRowsOpImpl

- This refers to compilation/linker errors that were observed in the
  github CI.

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+      continue;
+    }
+    GpuAtomicAdd(&elements_per_row[row], 1);
+    if (i > 0 && row < indices[(i - 1) * rank]) {
```
So you don't need to compare columns because that's the "row major" invariant?  If so, please add a comment.

benbarsdell replied:
That's correct, comment added.

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+// true if the indices are not ordered by row.
+template <typename Tindex>
+__global__ __launch_bounds__(1024) void CountElementsPerRowKernel(
+    GpuLaunchConfig cfg, Tindex dense_rows, int rank, const Tindex* indices,
```
Is `rank` always `2`?

benbarsdell replied:
No, it can be higher for higher-rank sparse tensors.

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+  GPU_1D_KERNEL_LOOP(i, cfg.virtual_thread_count) {
+    Tindex input_i = input_index_map ? input_index_map[i] : i;
+    Tindex row = indices[input_i * rank];
+    if (row < 0 || row >= dense_rows) {
+      continue;  // Ignore indices that are out of range
+    }
```
Is this needed?  Won't we have returned an error if any indices were out of range?

benbarsdell replied:
Good point, removed.

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+         num_empty_rows_through_t, num_empty_rows_through, input_row_ends_t,
+         input_row_ends, empty_row_indicator_t, empty_row_indicator,
+         done]() -> void {
+      CHECK(done);  // Crash OK
```
Can you please proactively change these to be `DCHECK`s to avoid accidentally holding up the merge?

benbarsdell replied:
Done.

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+      Tindex* input_index_map = nullptr;
+      Tensor input_index_map_t;
+      int rows_are_not_ordered = *rows_are_not_ordered_host.data();
+      if (rows_are_not_ordered) {
```
Can some of this logic be split into separate named functions?  I think that would make the flow more readable.

benbarsdell replied:
I've pulled the output allocations and sorting logic out into helper methods so the lambda is a lot shorter now.

sanjoy reviewed (changes requested):
One general comment: can you please refactor out some helper functions?  That will help readability I think.

benbarsdell committed:
Address minor comments from PR 46172

- Add comment re ordering by rows only.
- Remove unnecessary check for invalid indices (they are already checked
  beforehand).
- Change CHECK to DCHECK.

benbarsdell committed:
Refactor some SparseFillEmptyRows GPU logic

- Moves output allocations into a helper method.
- Moves sorting by rows logic into a helper method.

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+        index_type, TensorShape({dense_rows}), &num_empty_rows_through_t));
+    auto num_empty_rows_through = num_empty_rows_through_t.flat<Tindex>();
+
+    err = gpuprim::DeviceScan::InclusiveSum(
+        /*d_temp_storage=*/nullptr,
+        /*temp_storage_bytes=*/temp_storage_bytes,
+        /*d_in=*/empty_row_indicator,
+        /*d_out=*/num_empty_rows_through.data(),
+        /*num_items=*/dense_rows,
+        /*stream=*/cu_stream);
+    if (err != 0) {
+      return errors::Internal(
+          "SparseFillEmptyRows: Could not launch "
+          "gpuprim::DeviceScan::ExclusiveSum to calculate temp_storage_bytes, "
+          "status: ",
+          cudaGetErrorString(err));
+    }
+    {
+      Tensor temp_storage2;
+      TF_RETURN_IF_ERROR(context->allocate_temp(
+          DT_INT8, TensorShape({static_cast<int64>(temp_storage_bytes)}),
+          &temp_storage2));
+      err = gpuprim::DeviceScan::InclusiveSum(
+          /*d_temp_storage=*/temp_storage2.flat<int8>().data(),
+          /*temp_storage_bytes=*/temp_storage_bytes,
+          /*d_in=*/empty_row_indicator,
+          /*d_out=*/num_empty_rows_through.data(),
+          /*num_items=*/dense_rows,
+          /*stream=*/cu_stream);
+      if (err != 0) {
+        return errors::Internal(
+            "SparseFillEmptyRows: Could not launch "
+            "gpuprim::DeviceScan::ExclusiveSum to scan empty_row_indicator, "
+            "temp_storage_bytes:",
+            temp_storage_bytes, ", status: ", cudaGetErrorString(err));
+      }
+    }
```
This routine (first get temp bytes, allocate temp byte and do actual work) can be a helper function.

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+    if (err != 0) {
+      return errors::Internal(
+          "SparseFillEmptyRows: Could not launch "
+          "gpuprim::DeviceScan::ExclusiveSum to calculate temp_storage_bytes, "
```
InclusiveSum

sanjoy commented on tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:
```diff
@@ -0,0 +1,674 @@
+        keys_out_ptr, indices_in_ptr, indices_out_ptr, N, 0, sizeof(Tindex) * 8,
+        cu_stream);
+    return Status::OK();
+  }  // At this point cub_temp_storage will be marked for deallocation.
```
Misplaced comment?

sanjoy reviewed (changes requested):
Couple more cleanup suggestions.

benbarsdell committed:
Refactor SparseFillEmptyRows GPU to use gpu_prim_helpers

benbarsdell commented:
I've cleaned this up in https://github.com/tensorflow/tensorflow/pull/46172/commits/78353b59aa7f0b2bc88ad50f642a8044c1201468 using `gpu_prim_helpers.h` from the recently-merged https://github.com/tensorflow/tensorflow/pull/47082/. It addresses the three requested changes.

sanjoy commented:
Ben, this PR looks functionally correct to me, but I believe these ops only work on sparse matrices, which means `rank` is always 2.  If you agree, do you mind sending a separate PR simplifying the kernel and adding the runtime check?

CC @penpornk @ebrevdo

benbarsdell commented:
Thanks Sanjoy.

Regarding rank, the CPU implementation does currently support rank>2, and it is possible to pass a rank>2 `SparseTensor` to `sparse_fill_empty_rows` without getting any errors. The only indication I see of it being limited to rank=2 is this message in the documentation:`A SparseTensor with shape [N, M].`

Example with rank=3:
```
tf.sparse.fill_empty_rows(tf.sparse.SparseTensor(indices=[[0, 1, 2]], values=[3.14], dense_shape=[3, 4, 5]), default_value=1.23)
```

sanjoy commented:
> Thanks Sanjoy.
> 
> Regarding rank, the CPU implementation does currently support rank>2, and it is possible to pass a rank>2 `SparseTensor` to `sparse_fill_empty_rows` without getting any errors. The only indication I see of it being limited to rank=2 is this message in the documentation:`A SparseTensor with shape [N, M].`

I suspect that's just a missing runtime check.  Otherwise the spec of the op (which talks in terms of "rows") does not make sense to me.

> 
> Example with rank=3:
> 
> ```
> tf.sparse.fill_empty_rows(tf.sparse.SparseTensor(indices=[[0, 1, 2]], values=[3.14], dense_shape=[3, 4, 5]), default_value=1.23)
> ```

gbaned commented:
@benbarsdell  Can you please check @sanjoy's comments and keep us posted ? Thanks!

benbarsdell commented:
Any info on why this was reverted?
